tez-tests/src/test/java/org/apache/tez/test/TestAMRecoveryAggregationBroadcast.java

```diff
@@ -0,0 +1,513 @@
+        .build()
+        .createDefaultBroadcastEdgeProperty();
+
+    DAG dag = DAG.create("TestAMRecoveryAggregationBroadcast_" + dagName);
+    dag.addVertex(tableScanVertex)
+        .addVertex(aggregationVertex)
+        .addVertex(mapJoinVertex)
+        .addEdge(Edge.create(tableScanVertex, aggregationVertex, orderedEdge))
+        .addEdge(Edge.create(aggregationVertex, mapJoinVertex, broadcastEdge));
+    return dag;
+  }
+
+  TezCounters runDAGAndVerify(DAG dag, boolean killAM) throws Exception {
+    tezSession.waitTillReady();
+    DAGClient dagClient = tezSession.submitDAG(dag);
+
+    if (killAM) {
+      TimeUnit.SECONDS.sleep(10);
+      YarnClient yarnClient = YarnClient.createYarnClient();
+      yarnClient.init(tezConf);
+      yarnClient.start();
+      ApplicationAttemptId id = ApplicationAttemptId.newInstance(
+          tezSession.getAppMasterApplicationId(), 1);
+      yarnClient.failApplicationAttempt(id);
+      yarnClient.close();
+    }
+    DAGStatus dagStatus = dagClient
+        .waitForCompletionWithStatusUpdates(EnumSet.of(StatusGetOpts.GET_COUNTERS));
+    LOG.info("Diagnosis: " + dagStatus.getDiagnostics());
+    Assert.assertEquals(State.SUCCEEDED, dagStatus.getState());
+
+    FSDataInputStream in = remoteFs.open(new Path(OUT_PATH, "part-v002-o000-r-00000"));
+    ByteBuffer buf = ByteBuffer.allocate(100);
+    in.read(buf);
+    buf.flip();
+    Assert.assertEquals(EXPECTED_OUTPUT, StandardCharsets.UTF_8.decode(buf).toString());
+    return dagStatus.getDAGCounters();
+  }
+
+  private List<HistoryEvent> readRecoveryLog(int attemptNum) throws IOException {
+    ApplicationId appId = tezSession.getAppMasterApplicationId();
+    Path tezSystemStagingDir = TezCommonUtils.getTezSystemStagingPath(tezConf, appId.toString());
+    Path recoveryDataDir = TezCommonUtils.getRecoveryPath(tezSystemStagingDir, tezConf);
+    FileSystem fs = tezSystemStagingDir.getFileSystem(tezConf);
+    List<HistoryEvent> historyEvents = new ArrayList<>();
+    Path currentAttemptRecoveryDataDir = TezCommonUtils.getAttemptRecoveryPath(recoveryDataDir, attemptNum);
+    Path recoveryFilePath =
+        new Path(currentAttemptRecoveryDataDir, appId.toString().replace("application", "dag")
+            + "_1" + TezConstants.DAG_RECOVERY_RECOVER_FILE_SUFFIX);
+    if (fs.exists(recoveryFilePath)) {
+      LOG.info("Read recovery file:" + recoveryFilePath);
+      historyEvents.addAll(RecoveryParser.parseDAGRecoveryFile(fs.open(recoveryFilePath)));
+    }
+    printHistoryEvents(historyEvents, attemptNum);
+    return historyEvents;
+  }
+
+  private void printHistoryEvents(List<HistoryEvent> historyEvents, int attemptId) {
+    LOG.info("RecoveryLogs from attempt:" + attemptId);
+    for(HistoryEvent historyEvent : historyEvents) {
+      LOG.info("Parsed event from recovery stream"
+          + ", eventType=" + historyEvent.getEventType()
+          + ", event=" + historyEvent);
+    }
+    LOG.info("");
```

Comment:
hm, this looks strange, is this really needed? printed only once after all the events are printed

Reply:
I ported it from [TestAMRecovery](https://github.com/apache/tez/blob/1084699ddb1baf6e4a6df97d988ac696d33f41e8/tez-tests/src/test/java/org/apache/tez/test/TestAMRecovery.java#L506). I guess we wanted a new line to visibly separate huge output.